File: src/Browser/Avalonia.Browser/webapp/modules/avalonia/navigationHelper.ts
export class NavigationHelper {
    public static addBackHandler(backHandlerCallback: () => Boolean) {
        history.pushState(null, "", window.location.href);
        window.onpopstate = () => {
            const handled = backHandlerCallback();

            if (!handled) {
                history.back();
            } else {
                history.forward();
            }
        };
    }

    public static openUri(uri?: string, target?: string) {
        return !!window.open(uri, target);
    }
}

File: src/Browser/Avalonia.Browser/Interop/NavigationHelper.cs
using System;
using System.Runtime.InteropServices.JavaScript;
using System.Threading.Tasks;
using Avalonia.Platform;

namespace Avalonia.Browser.Interop;

internal static partial class NavigationHelper
{
    [JSImport("NavigationHelper.addBackHandler", AvaloniaModule.MainModuleName)]
    public static partial void AddBackHandler([JSMarshalAs<JSType.Function<JSType.Boolean>>] Func<bool> backHandlerCallback);

    public static Task<bool> OnBackRequested()
    {
        var handled = (AvaloniaLocator.Current.GetService<ISystemNavigationManagerImpl>() as BrowserSystemNavigationManagerImpl)?
            .OnBackRequested() ?? false;
        return Task.FromResult(handled);
    }

    [JSImport("NavigationHelper.openUri", AvaloniaModule.MainModuleName)]
    public static partial bool WindowOpen(string uri, string target);
}
